core/src/ast/macros.rs

```diff
@@ -0,0 +1,302 @@
+use std::{collections::BTreeMap, fmt::Debug};
+
+use proc_macro2::{TokenStream, TokenTree};
+use quote::{ToTokens, TokenStreamExt};
+use syn::{
+    braced, bracketed,
+    buffer::{Cursor, TokenBuffer},
+    parenthesized,
+    parse::{self, Parse},
+    token, Error, Expr, Ident, ImplItem, ImplItemMacro, Item, ItemMacro, Token,
+};
+
+#[derive(Default)]
+pub struct Macros {
+    defs: BTreeMap<Ident, MacroRules>,
+}
+
+impl Macros {
+    pub fn new() -> Macros {
+        Macros {
+            defs: BTreeMap::new(),
+        }
+    }
+
+    pub fn read_item_macro(&mut self, input: &ItemMacro) -> Option<Vec<Item>> {
+        let mac = Macro::from_syn(input);
+        if let Ok((ident, mac)) = mac {
+            match mac {
+                Macro::MacroRules(rules) => {
+                    self.defs.insert(ident, rules);
+                    // Macro rules add no new items:
+                    None
+                }
+                Macro::MacroMatch(matched) => {
+                    if let Some(def) = self.defs.get(&ident) {
+                        Some(def.evaluate(matched))
+                    } else {
+                        panic!("Could not find definition for {ident}. Have you tried creating a #[diplomat::macro_rules] macro_rules! {ident} definition?");
+                    }
+                }
+            }
+        } else {
+            // We handle errors automatically in `diplomat/macro`
+            None
+        }
+    }
+
+    pub fn read_impl_item_macro(&self, input: &ImplItemMacro) -> Vec<ImplItem> {
+        let m: syn::Result<MacroMatch> = input.mac.parse_body();
+        // FIXME: Extremely hacky. In the future for importing macros, we'll want to do something else.
+        let path_ident = input.mac.path.segments.last().unwrap().ident.clone();
+
+        if let Ok(matched) = m {
+            if let Some(def) = self.defs.get(&path_ident) {
+                def.evaluate(matched)
+            } else {
+                panic!("Could not find definition for {path_ident}. Have you tried creating a #[diplomat::macro_rules] macro_rules! {path_ident} definition?");
+            }
+        } else {
+            // We handle errors automatically in `diplomat/macro`
+            Vec::new()
+        }
+    }
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Macro {
```

Comment:
issue: what's the benefit of treating these as an enum? they should be constructed and handled in different places. I think we should have `MacroDef` / `MacroUse` or something similar to everything else.